File: sword_offer/cpp/35_数组中的逆序对.cpp
#include <iostream>
#include <vector>
#include <ctime>
#include <cstdlib>
#include <stdio.h>
#include <sys/time.h>

class Solution {
public:
    // 暴力枚举
    int InversePairs1(std::vector<int> nums) {
        if (nums.empty()) {
            return 0;
        }

        int count = 0;
        for (int i = 0; i < nums.size() - 1; i++) {
            for (int j = i + 1; j < nums.size(); j++) {
                if (nums[i] > nums[j]) {
                    count++;
                }
            }
        }

        return count;
    }

    // 高效方式：归并排序
    long cnt = 0;
    std::vector<int> tmp;
    int InversePairs(std::vector<int> nums) {
        if (nums.empty()) {
            return 0;
        }
        
        tmp = nums;
        mergeSort(nums, 0, nums.size() - 1);
        return int(cnt % 1000000007);
    }

    // 递归调用的函数
    // 注意使用引用，否则排序后无法替换原数组的值
    // 这里其实是二分的思想
    void mergeSort(std::vector<int>& nums, int l, int h) {
        if (l >= h) {
            // 当分组到只剩一个时，返回上一层，开始merge
            return;
        }
        int m = l + (h - l) / 2;
        mergeSort(nums, l, m);     // 左数组，统计，排序
        mergeSort(nums, m + 1, h); // 右数组，统计，排序
        merge(nums, l, m, h);      // 本次的前后数组，统计，排序
    }
    
    // 用来实现当前前后数组统计与合并的函数
    // 将两个有序数组合并为一个，同时统计逆序对
    // 注意使用引用
    void merge(std::vector<int>& nums, int l, int m, int h) {
        int i = l, j = m + 1, k = l;
        while (i <= m || j <= h) {
            if (i > m) { // 此时前数组已被遍历完，所有前数组小于后数组当前值
                tmp[k++] = nums[j++];
            } else if (j > h) { // 后数组被遍历完
                tmp[k++] = nums[i++];
            } else if (nums[i] <= nums[j]) { // 两个数组都没有遍历完，前 < 后
                tmp[k++] = nums[i++];
            } else { // 两个数组都没有遍历完，前 > 后，符合逆序数要求
                tmp[k++] = nums[j++];
                this->cnt += m - i + 1;
            }
        }
        
        // 将完成排序的部分拷贝到原数组中
        for (k = l; k <= h; k++) {
            nums[k] = tmp[k];
        }
    }

    // [NOTE](lzc/2020/10/07):改进版归并合并函数
	void merge(std::vector<int>& nums, int l, int m, int h) {
		int* tmp = new int[h - l + 1];
		for (int i = l; i <= h; i++) {
			tmp[i] = nums[i - l];
		}
		int p = l;
		int q = m + 1;
		for (int i = l; i <= h; i++) {
			if (p <= m && q >= h && tmp[p - l] > tmp[q - l]) {
				cnt += m - q + 1;
			}
			if (p <= m && (q > h || tmp[p - l] <= tmp[q - l])) {
				nums[i] = tmp[p - l];
				p += 1;
			} else {
				nums[i] = tmp[q - l];
				q += 1;
			}
		}
		return;
	}

    // 测试函数
    void test(std::vector<int>& nums) {
        int result = 0;
        struct timeval start, end;
        for (auto func : this->func_vec_) {
            // 必要的处理，防止后续调用干扰
            this->cnt = 0;

            // 正式调用
            gettimeofday(&start, 0);
            result = (this->*func)(nums);
            gettimeofday(&end, 0);
            printf("result: %d, time(us): %ld\n", result, end.tv_usec - start.tv_usec);
        }
    }

private:
    typedef int (Solution::*func_ptr)(std::vector<int>);
    std::vector<func_ptr> func_vec_ = {&Solution::InversePairs1, &Solution::InversePairs};
};

// 获取指定数目和范围的随机数组
void get_rand_vec(std::vector<int>& nums, int range, int length) {
    if (range <= 0) {
        return;
    }

    srand((unsigned)time(nullptr)); // 设置随机数种子
    for (int i = 0; i < length; i++) {
        nums.push_back(rand() % range);
    }
}

int main(int argc, char* argv[])
{
    std::vector<int> nums = {1, 2, 3, 4, 5, 6, 7, 0};
    std::vector<int> nums2;
    get_rand_vec(nums2, 100, 100);
 
    Solution s;
    s.test(nums);
    std::cout << "=====" << std::endl;
    s.test(nums2);

    return 0;
}

File: sword_offer/python/35_数组中的逆序对.py
#!/bin/bash python3
#-*- coding:utf-8 -*-

import time
import random

class Solution():
    def __init__(self):
        self.cnt = 0
    
    def InversePairs1(self, nums):
        """
        暴力枚举
        """
        if nums == []:
            return 0
        
        count = 0
        for i in range(0, len(nums) - 1):
            for j in range(i + 1, len(nums)):
                if nums[i] > nums[j]:
                    count += 1
        
        return count

    def InversePairs(self, nums):
        """
        注意，这种方式没有对外界的nums进行保护
        """
        if nums == []:
            return 0
        self.mergeSort(nums, 0, len(nums) - 1)
        return self.cnt % 1000000007
        
    def mergeSort(self, nums, l, h):
        if l >= h:
            return
        m = l + (h - l) // 2
        self.mergeSort(nums, l, m)
        self.mergeSort(nums, m + 1, h)
        self.merge(nums, l, m, h)

    def merge(self, nums, l, m, h):
        i = l
        j = m + 1
        k = l
        tmp = []
        while i <= m or j <= h:
            if i > m: # 前数组已经遍历完
                tmp.append(nums[j])
                j += 1
            elif j > h or nums[i] <= nums[j]: # 后数组已经遍历完，或则　前 < 后
                tmp.append(nums[i])
                i += 1
            else: # 前 > 后，符合逆序对的定义
                tmp.append(nums[j])
                j += 1
                self.cnt += m - i + 1
            k += 1

        # 拷贝数组
        nums[l:h+1] = tmp[:]
    
    def merge2(self, nums, l, m, h):
        """
        [NOTE](lzc/2020/10/07):改进版归并合并函数
        """
        tmp = nums[l:h+1]
        p, q = l, m + 1
        for i in range(l, h + 1):
            if p <= m and q <= h and tmp[p - l] > tmp[q - l]:
                self.cnt += m - q + 1
            if p <= m and (q > h or tmp[p - l] <= tmp[q - l]):
                nums[i] = tmp[p - l]
                p += 1
            else:
                nums[i] = tmp[q - l]
                q += 1
        return

    def test(self, nums):
        """
        测试函数
        """
        func_vec = [self.InversePairs1, self.InversePairs]
        for func in func_vec:
            # 必要的处理，防止后续调用影响
            tmp_nums = nums[:]
            self.cnt = 0

            # 正式调用
            start = time.time()
            result = func(tmp_nums)
            end = time.time()
            print("result: {:d}, time(us): {:>5.2f}".format(result, (end - start)*10**6))


def main():
    nums = [1, 2, 3, 4, 5, 6, 7, 0]
    nums2 = random.sample(range(100), 100) # 这种方式生成0-100的乱序

    s = Solution()
    s.test(nums)
    print("=====")
    s.test(nums2)


if __name__ == "__main__":
    main()